drivers/i2c/i2c_nrfx_twi_rtio.c

```diff
@@ -0,0 +1,365 @@
+/*
+ * Copyright (c) 2018, Nordic Semiconductor ASA
+ * Copyright (c) 2024, Croxel Inc
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/i2c/rtio.h>
+#include <zephyr/dt-bindings/i2c/i2c.h>
+#include <zephyr/pm/device.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <soc.h>
+#include <nrfx_twi.h>
+
+#include <zephyr/logging/log.h>
+#include <zephyr/irq.h>
+LOG_MODULE_REGISTER(i2c_nrfx_twi, CONFIG_I2C_LOG_LEVEL);
+
+#if CONFIG_I2C_NRFX_TRANSFER_TIMEOUT
+#define I2C_TRANSFER_TIMEOUT_MSEC K_MSEC(CONFIG_I2C_NRFX_TRANSFER_TIMEOUT)
+#else
+#define I2C_TRANSFER_TIMEOUT_MSEC K_FOREVER
+#endif
+
+struct i2c_nrfx_twi_data {
+	struct i2c_rtio *ctx;
+	struct k_sem transfer_sync;
+	struct k_sem completion_sync;
+	volatile nrfx_err_t res;
+	uint32_t dev_config;
+};
+
+struct i2c_nrfx_twi_config {
+	nrfx_twi_t twi;
+	nrfx_twi_config_t config;
+	const struct pinctrl_dev_config *pcfg;
+};
+
+static int i2c_nrfx_twi_recover_bus(const struct device *dev);
+static void i2c_nrfx_twi_complete(const struct device *dev, int status);
+
+static bool i2c_nrfx_twi_msg_start(const struct device *dev, uint8_t flags,
+				   uint8_t *buf, size_t buf_len, uint16_t i2c_addr)
+{
+	const struct i2c_nrfx_twi_config *config = dev->config;
+	struct i2c_nrfx_twi_data *const dev_data = dev->data;
+	struct i2c_rtio *ctx = dev_data->ctx;
+	int ret = 0;
+
+	/** Enabling while already enabled ends up in a failed assertion: skip it. */
+	if (k_sem_take(&dev_data->completion_sync, K_NO_WAIT) == 0) {
```

Comment:
start should only be called if nothing else is in progress, is this semaphore really needed?

Reply:
The logic followed in this code attempts to replicate this sequence: https://github.com/zephyrproject-rtos/zephyr/blob/2c6306d09954de13f993ebc28e173246b9b6ef39/drivers/i2c/i2c_nrfx_twi.c#L55

consisted of:
1. Enable the TWI periheral before starting the set of messages.
2. Go through each message, transferring one after the other.
3. Disable the TWI peripheral.

In this driver, `i2c_nrfx_twi_msg_start()` is called on every message, and the semaphore signals whether it's the beginning of the set of messages.

LMK if you think we should structure it differently.

Reply:
Ah, you could use a simple boolean flag I believe in this case as a data member, semaphore are lightweight but a boolean seems sufficient in this case, no threads are going to be waiting on this semaphore

Reply:
You're right - I've changed this to a boolean named `twi_enabled`.